PolyEngine/Core/Src/Optional.hpp

```diff
@@ -0,0 +1,188 @@
+				this->Value() = std::move(other.Value());
+
+				other.Value().~V();
+				other.initialized = false;
+			} else if (this->HasValue()) {
+				this->Value().~V();
+				this->initialized = false;
+			}
+			return *this;
+		}
+
+	public:
+		bool HasValue() const { return this->initialized; }
+		operator bool() const { return this->HasValue(); }
+
+	public:
+		const V&  Value() const& { ASSERTE(this->HasValue(), "Attempting to get a value from an empty optional"); return this->value_storage; }
+		      V&  Value() &      { ASSERTE(this->HasValue(), "Attempting to get a value from an empty optional"); return this->value_storage; }
+		      V&& Value() &&     { ASSERTE(this->HasValue(), "Attempting to get a value from an empty optional"); return std::move(this->value_storage); }
+
+	public:
+		V TakeValue() {
+			ASSERTE(this->HasValue(), "Attempting to take a value from an empty optional");
+
+			V ret = std::move(this->value_storage);
+
+			this->initialized = false;
+			this->value_storage.~V();
+
+			return ret;
+		}
+
+	public:
+		V ValueOr(V&& fallback) const &  { if (this->HasValue()) { return this->Value(); } return fallback; } //todo(vuko): do not copy when possible
+		V ValueOr(V&& fallback) const && { if (this->HasValue()) { return std::move(this->Value()); } return fallback; }
+
+	private:
+		bool initialized;
+		union {
+			V value_storage;
+			char dummy;
+		};
+	};
+
+	template<typename V>
+	class Optional<V&> {
+	public:
+		using ValueType = V&;
+	public:
+		Optional() : value_storage(nullptr) {}
+		Optional(V& value) : value_storage(&value) {}
+		~Optional() = default;
+
+	public:
+		Optional(Optional&& other) { this->value_storage = other.value_storage; other.value_storage = nullptr; }
+		Optional& operator=(Optional&& other) { this->value_storage = other.value_storage; other.value_storage = nullptr; return *this; }
+
+	public:
+		bool HasValue() const  { return static_cast<bool>(this->value_storage); }
+		operator bool() const { return this->HasValue(); }
+
+	public:
+		const V&  Value() const& { ASSERTE(this->HasValue(), "Attempting to get a value from an empty optional"); return *this->value_storage; }
+		      V&  Value() &      { ASSERTE(this->HasValue(), "Attempting to get a value from an empty optional"); return *this->value_storage; }
+
+	public:
+		V& TakeValue() { ASSERTE(this->HasValue(), "Attempting to take a value from an empty optional"); V& ret = *this->Value(); this->value_storage = nullptr; return ret; }
+
+	public:
+		V ValueOr(V&& fallback) const &  { if (this->HasValue()) { return *this->Value(); } return fallback; }
+		V ValueOr(V&& fallback) const && { if (this->HasValue()) { return std::move(*this->Value()); } return fallback; }
+
+	private:
+		V* value_storage;
+	};
+
+	//todo(vuko): this is pretty much copy-pasted from the main implementation (with the copying constructor removed); can we somehow make an alias?
+	//always moves in the contents
+	template<typename V>
+	class Optional<V&&> {
```

Comment:
It's a convenience specialization. `Optional<T&&>` is effectively the same as `Optional<T>`, but it forbids passing in lvalue references. Another thing to note is that the specialization makes
Optionals play nicer with forwarding references. Consider this:
```
template<typename T>
void func(T&& t) {
	Optional<T>(std::forward<T>(t));
	//...do stuff with the optional
}
```
When passing a lvalue reference, no copying takes place, yet rvalue references are moved in as expected.